[6.0][PackageCMO] Don't allow modifying AST

elsh opened the pull request:
This PR prevents @usableFromInline attribute from being added to an internal type, which modifies AST and should not be allowed in PackageCMO. Also ensures that the generated interfaces with PackageCMO enabled are not affected by the optimization and do not contain modified AST.

rdar://130292190

main: https://github.com/swiftlang/swift/pull/74641

elsh commented:
@swift-ci test

tbkka commented:
Looks like the main PR still has a "requested changes" review from @eeckstein.  You should address that before cherry-picking to the 6.0 branch.

elsh committed:
[PackageCMO] Don't allow modifying AST.

 Currently not all types are visited in canSerialize* calls, sometimes
 resulting in an internal type getting @usableFromInline, which is
 incorrect.

 For example, for `let q = P() as? Q`, where Q is an internal class
 inherting a public class P, Q is not visited in the canSerialize*
 checks, thus resulting in `@usableFromInline class Q`; this is not
 the intended behavior in the conservative mode used by PackageCMO
 as it modifies AST.

 To properly fix, instruction visitor needs to be refactored to do
 both the "canSerialize" check (that visits all types) and serialize
 or update visibility (modify AST in non-conservative modes).

 This PR provides a short-term fix that prevents modifying AST, and
 also ensures that the generated interfaces with PackageCMO flags
 are not affected by the optimization or contain modified AST.

rdar://130292190

elsh commented:
@swift-ci test

elsh commented:
@swift-ci test

elsh commented:
@swift-ci test
